pkgs/build-support/setup-hooks/auto-patchelf.py

```diff
@@ -0,0 +1,304 @@
+
+    for dep in missing:
+        if dep.is_absolute():
+            # This is an absolute path. If it exists, just use it. Otherwise,
+            # we probably want this to produce an error when checked (because
+            # just updating the rpath won't satisfy it).
+            if dep.is_file():
+                continue
+        elif os.path.isfile(os.path.join(libcLib, dep)):
+            # This library exists in libc, and will be correctly resolved by
+            # the linker.
+            continue
+
+        if foundDependency := findDependency(dep, fileArch, fileOsabi):
+            rpath.append(foundDependency)
+            print("    {} -> found: {}".format(dep, foundDependency))
+        else:
+            print("    {} -> not found!".format(dep))
+            yield dep
+
+    # dedup. Relies on dicts preserving insertion order. Python3.7+
+    rpath = ":".join(dict.fromkeys(rpath))
+
+    if rpath:
+        print("setting RPATH to:", rpath)
+        subprocess.run(["patchelf", "--set-rpath", rpath, elf.filename], check = True)
+
+
+def walk(paths, recurse):
+    for path in paths:
+        for (dirpath, dirs, files) in os.walk(path):
+            for file in files:
+                fullPath = os.path.join(dirpath, file) # todo: normalize
+                yield fullPath
+
+            if not recurse:
+                break
+
+
+def autoPatchelf(pathsToPatch, libDirs, runtimeDeps, recurse = True, ignoreMissing = False):
+    if not pathsToPatch:
+        sys.exit("no paths to patch, stopping.")
+
+    # Add all shared objects of the current output path to the cache,
+    # before libDirs, so that they are chosen first in findDependency.
+    populateCache(pathsToPatch, recurse)
+    populateCache(libDirs)
+
+    missingDeps = {}
+    for path in chain.from_iterable(p.rglob('*') if recurse else p.glob('*')
+                                    for p in pathsToPatch):
+        if path.is_dir():
+            continue
+        try:
+            with openELF(path) as elf:
+                missing = list(autoPatchelfFile(elf, runtimeDeps))
+                if missing:
+                    missingDeps[path] = missing
+        except ELFError:
+            pass
+
+    # Print a summary of the missing dependencies at the end
+    for path, deps in missingDeps.items():
+        for dep in deps:
+            print(f"autoPatchelfHook could not satisfy dependency {dep} wanted by {path}")
+
+    if missingDeps and not ignoreMissing:
+        sys.exit("""
+            autoPatchelfHook failed to find all the required dependencies.
+            Add the missing dependencies to the build inputs or set autoPatchelfIgnoreMissingDeps=true.
+            """)
+
+def main():
+
+    parser = argparse.ArgumentParser(prog="autoPatchelf.py",
+            description="""autoPatchelf tries as hard as possible to patch the provided\
+                    binary files by looking for compatible libraries in the provided paths.
+                    """)
```

Comment:
Also nitpick, the following would avoid the unnecessary whitespace:

```suggestion
        description='Patch automatically discovered library dependencies into'
                    ' the provided executable.')
```